### ISSUE TITLE
Upstream this flake

### ISSUE THREAD

**Issue by nakibrayan3** [sodiboo/niri-flake#480]
Hello @sodiboo , thank you so much for your work, this flake is amazing.

Is possible to upstream nixosModules, homeManagerModules and stylixModules to their respective repos, so that it's easier to user niri in nixos.

**Comment by sodiboo** [comment]
huh. turns out i have not replied at all to this. yes, eventually i'd love for most of it to be upstreamed

Yeah, so i'm gonna use this as a tracking issue for general upstream work:

- We now have the core functionality off the nixos module in nixpkgs (https://github.com/NixOS/nixpkgs/pull/348193)
- The NixOS module has no configuration ability (e.g. `/etc/niri/config.kdl`)

There are two parts to configuration, firstly the `config` option to set the verbatim config syntax, and because niri's config is an XML-ish language rather than a JSON-ish language, it's difficult to just use in Nix. As such, the `settings` option from this flake is a mapping of the niri config onto a JSON-ish data model.

- I plan to open a PR to nixpkgs to add this `config` style of option. It would potentially be blocked by https://github.com/NixOS/nixpkgs/pull/295211 if we don't want to change the type retroactively. But realistically, we absolutely can do so and it should be a string until that PR is merged.
- I don't foresee the `settings` option ever being upstreamed to nixpkgs. It is very package-version-dependent, and will probably already remain in niri-flake. Maybe at some point this will be the main thing i provide.

---

- The "main" home-manager module (`niri.homeModules.niri`) is honestly janky, and difficult to use **correctly**, so i don't actually think it ought to be upstreamed. In particular, standalone home-manager has no way to ensure the mesa version is correct, and nixos-based home-manager doesn't benefit from such a module. home-manager should probably only have a config module
- Again, like in NixOS, this is somewhat dependent on KDL support in nixpkgs. Like, we can't just add [this](https://github.com/sodiboo/niri-flake/blob/main/kdl.nix) to home-manager next to niri, because when the nixpkgs KDL support is merged, that *should* provide the type, and we definitely don't want to switch to a different KDL type.
- Again, like in NixOS, i don't really foresee the `settings` module being upstreamed to home-manager. Because it will need to be updated for unstable commits as well.

---

The only way a `settings` option can be reasonably upstreamed is if niri supports merging partial configs. That way, we can have an incomplete `settings` and that's okay if we also add `extraConfig` as an option to append to the config file. Currently, niri doesn't support any such thing.

---

A stylix module is 100% dependent on a reasonable nix-based `settings` option. It doesn't work without that. So, this will likely never be upstreamed without changes to niri to allow for partial config parsing and merging (because otherwise, we will not have an upstream structured `settings` option)

**Comment by cgahr** [comment]
Is it necessary to expose all possible niri config options?

For example, it is possible to configure sway via [home-manager](https://github.com/nix-community/home-manager/blob/master/modules/services/window-managers/i3-sway/sway.nix).
The module exposes many important options but not all possible config options. Instead, the sway module focuses on the most important and ergonomic config options. The missing ones can always be added via `extraConfig` as plain text.

My favorite example is the option to change the default movement keys from `hjkl` to something else. IMO from a user perspective, this is great since it hides the boilerplate of renaming all variants of `hjkl`.

`stylix` could work similarly: expose the 3-4 important coloring options (for example omit the gradient coloring of border), and stylix sets those colors automatically and merges them.

This way, it is possible to configure the most important parts via home-manager, everything else is only necessary for few users and has to be done by hand in `extraConfig`.

What do you think?

**Comment by cgahr** [comment]
Is merging configs necessary? 

I imagine the following nix structure:
```nix
{
  input = [
    {
      keyboard = {
        xkb = { };
      };
    }
    {
      touchpad = { };
    }
  ];

  # maybe outputs should be a list
  outputs = {
    "eDP-1" = {
      scale = 1;
    };
  };

  layout = {
    gaps = 0;
    preset-column-widths = ''
      proportion 0.33333
      proportion 0.5
      proportion 0.66667
    '';

    struts = {
      left = 64;
      right = 64;
    };
  };

  screenshot-path = ~/.cache/screenshots;

  window-rules = [
    # maybe just strings?
  ];

  binds = {
    "Mod+Left" = "focus-column-left";
    "Mod+Down" = "focus-window-down";
    "Mod+Up" = "focus-window-up";
    "Mod+Right" = "focus-column-right";
    "Mod+M" = "focus-column-left";
    "Mod+N" = "focus-window-down";
    "Mod+E" = "focus-window-up";
    "Mod+I" = "focus-column-right";

    # how to handle cooldown-ms?
    # and allow-inhibiting?
  };
  bindsExtraConfig = null;
}

```
As far as I can tell, most config options of niri are `key=value` pairs. They are easily encoded in nix.
The only options not matching this pattern are `preset-column-widths`, keybinds with `cooldown-ms or `allow-inhibiting` and (probably) window-rules. Those could just be pure string fields.

stylix uses normal nix syntax to set everything, and nix merges all config options normally. If necessary, there is `outputsExtraConfig`, `bindsExtraConfig` and others that add strings at the end. Would this work?

**Comment by Ericson2314** [comment]
Can we at least get the equivalent of the NixOS module in home manager?

**Comment by sodiboo** [comment]
> Can we at least get the equivalent of the NixOS module in home manager?

Because niri requires - for some reason - an exact mesa version match with the system, this would inherently break frequently on every distro.

**Comment by Ericson2314** [comment]
> niri requires - for some reason - an exact mesa version match with the system,

Huh, Is there a Niri issue for that?

**Comment by sodiboo** [comment]
> > niri requires - for some reason - an exact mesa version match with the system,
> 
> Huh, Is there a Niri issue for that?

I don't think so, because there's nothing for niri to do here. There isn't a good reason why niri should require this. It doesn't do anything weird. It's not niri's fault. It's really a nix issue, because NixOS graphics are *weird*. "requires an exact mesa version match" is probably not even the correct root cause, but matching them up exactly *does* seem to always fix the issue.

On NixOS, a "standalone home-manager installs niri" setup will break anytime your nixpkgs desyncs in home-manager vs operating system.
On other systems, niri just wouldn't work to begin with, unless you use [`nix-system-graphics`](https://github.com/soupglasses/nix-system-graphics) or similar.
